client/web/src/global/CoolCodeIntel.tsx

```diff
@@ -0,0 +1,842 @@
+                        start: {
+                            line: range.start.line + 1,
+                            character: range.start.character + 1,
+                        },
+                        end: { line: range.end.line + 1, character: range.end.character + 1 },
+                    },
+                })
+            )
+        )
+    }
+    return path
+}
+
+const LocationsList: React.FunctionComponent<{
+    locations: Location[]
+    activeLocation?: Location
+    setActiveLocation: (reference: Location | undefined) => void
+    filter: string | undefined
+}> = ({ locations, activeLocation, setActiveLocation, filter }) => {
+    const byFile: Record<string, Location[]> = {}
+    for (const location of locations) {
+        if (byFile[location.resource.path] === undefined) {
+            byFile[location.resource.path] = []
+        }
+        byFile[location.resource.path].push(location)
+    }
+
+    const locationGroups: LocationGroup[] = []
+    Object.keys(byFile).map(path => {
+        const references = byFile[path]
+        const repoName = references[0].resource.repository.name
+        locationGroups.push({ path, locations: references, repoName })
+    })
+
+    const byRepo: Record<string, LocationGroup[]> = {}
+    for (const group of locationGroups) {
+        if (byRepo[group.repoName] === undefined) {
+            byRepo[group.repoName] = []
+        }
+        byRepo[group.repoName].push(group)
+    }
+    const repoLocationGroups: RepoLocationGroup[] = []
+    Object.keys(byRepo).map(repoName => {
+        const referenceGroups = byRepo[repoName]
+        repoLocationGroups.push({ repoName, referenceGroups })
+    })
+
+    const getLineContent = (location: Location): string => {
+        const range = location.range
+        if (range !== undefined) {
+            return location.lines[range.start?.line].trim()
+        }
+        return ''
+    }
+
+    return (
+        <>
+            {repoLocationGroups.map(repoReferenceGroup => (
+                <RepoReferenceGroup
+                    key={repoReferenceGroup.repoName}
+                    repoReferenceGroup={repoReferenceGroup}
+                    activeLocation={activeLocation}
+                    setActiveLocation={setActiveLocation}
+                    getLineContent={getLineContent}
+                    filter={filter}
+                />
+            ))}
+        </>
+    )
+}
+
+const RepoReferenceGroup: React.FunctionComponent<{
+    repoReferenceGroup: RepoLocationGroup
+    activeLocation?: Location
+    setActiveLocation: (reference: Location | undefined) => void
+    getLineContent: (location: Location) => string
+    filter: string | undefined
+}> = ({ repoReferenceGroup, setActiveLocation, getLineContent, activeLocation, filter }) => {
+    const [isOpen, setOpen] = useState<boolean>(true)
+    const handleOpen = useCallback(() => setOpen(!isOpen), [isOpen])
```

Comment:
I think this would give us one less rerenders

```suggestion
    const handleOpen = useCallback(() => setOpen(previousState => !previousState), [])
```